### ISSUE TITLE
RTNETLINK errors

### ISSUE THREAD

**Issue by tongson** [zerotier/ZeroTierOne#994]
**Describe the bug**
Connection lost after a few minutes.
I see the following from the output of the 1.4.0 zerotier-one process:

```
Aug 06 07:13:21 zerotier-one: Error binding RTNETLINK (_requiestIPv4Routes #1): Address in use                                             
Aug 06 07:13:21 zerotier-one: Error binding RTNETLINK (_requestIPv6Routes #1): Address in use                                              
Aug 06 07:13:21 zerotier-one: Error binding RTNETLINK (_requestInterfaceList #1): Address in use
```

**Expected behavior**
When using 1.2.12 on the same hosts, no errors from the zerotier-one process and the connection is stable. Runs without problems for months.

 - OS: Linux
 - Ubuntu and openSUSE
 - ZeroTier Version 1.4.0

**Comment by tongson** [comment]
Additional info: I run zerotier-one across containers sharing the host network.

**Comment by tongson** [comment]
Additional info: We use our own root/planet/moon. The root is running 1.4.0 and 1.2.12 peers experienced no issues since the root upgrade to 1.4.0.

**Comment by zyclonite** [comment]
i see a similar behavior but it happens directly, creates a network interface but cannot join the network or assign an ip address

```
docker[5165]: Error binding RTNETLINK (addAddress #1): Address already in use
docker[5165]: Error binding RTNETLINK (_requestInterfaceList #1): Address already in use
docker[5165]: Error binding RTNETLINK (_requestIPv6Routes #1): Address already in use
docker[5165]: Error binding RTNETLINK (_requiestIPv4Routes #1): Address already in use
```

docker command
```docker run --device=/dev/net/tun --net=host --cap-add=NET_ADMIN --cap-add=SYS_ADMIN -v /var/lib/zerotier-one:/var/lib/zerotier-one zyclonite/zerotier:1.4.2```

* OS: Container Linux
* ZT Versions tested: 1.4.0, 1.4.2, 1.4.2-2
* dockerfile: https://github.com/znx/zerotier-docker
* docker containers: https://hub.docker.com/r/zyclonite/zerotier/tags

**Comment by adamierymenko** [comment]
This would have nothing to do with roots or anything else network related. This seems to be an issue with the ability of ZeroTier to make modifications to network interfaces such as assigning IPs.

Can you see if `/sbin/ip` can be used in these containers to assign IPs to these interfaces?

**Comment by adamierymenko** [comment]
It might also be very helpful to `strace` the ZeroTier binary and see exactly what systems calls are failing prior to these error outputs and what the system call failure return values and/or errno values are.

**Comment by zyclonite** [comment]
> This would have nothing to do with roots or anything else network related. This seems to be an issue with the ability of ZeroTier to make modifications to network interfaces such as assigning IPs.
> 
> Can you see if `/sbin/ip` can be used in these containers to assign IPs to these interfaces?

if it is used i cannot tell... but i can use it to assign ips from inside the container

after zt joins a network and is authorized the interface looks like this (maybe this helps)
```
ztr2qwaf6g: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 2800 qdisc fq_codel state UNKNOWN group default qlen 1000
    link/ether 32:66:ed:3d:0d:af brd ff:ff:ff:ff:ff:ff
    inet6 fe80::3066:edff:fe3d:daf/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by zyclonite** [comment]
standby a bit... i was able to get it joining a network and assigning an ip address on another node, i'll try to figure out why

the only difference i know... on the not working node are iptable rules in place

**Comment by zyclonite** [comment]
strace output on joining a network
```
socket(PF_NETLINK, SOCK_RAW, NETLINK_ROUTE) = 28
setsockopt(28, SOL_SOCKET, SO_RCVTIMEO, "\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
getpid()                                = 1
bind(28, {sa_family=AF_NETLINK, pid=1, groups=00000040}, 12) = -1 EADDRINUSE (Address already in use)
write(2, "Error binding RTNETLINK (_requie"..., 73) = 73
close(28)                               = 0
socket(PF_NETLINK, SOCK_RAW, NETLINK_ROUTE) = 28
setsockopt(28, SOL_SOCKET, SO_RCVTIMEO, "\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
getpid()                                = 1
bind(28, {sa_family=AF_NETLINK, pid=1, groups=00000400}, 12) = -1 EADDRINUSE (Address already in use)
write(2, "Error binding RTNETLINK (_reques"..., 72) = 72
close(28)                               = 0
socket(PF_NETLINK, SOCK_RAW, NETLINK_ROUTE) = 28
setsockopt(28, SOL_SOCKET, SO_RCVTIMEO, "\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
getpid()                                = 1
bind(28, {sa_family=AF_NETLINK, pid=1, groups=00000001}, 12) = -1 EADDRINUSE (Address already in use)
write(2, "Error binding RTNETLINK (_reques"..., 75) = 75
close(28)                               = 0
```

looks like getpid() does conflict with the host system (not unique), maybe setting it to 0 and let the kernel decide would fix it?

**Comment by zyclonite** [comment]
some more findings

it works with ubuntu 19.04 and docker 18.09.7
but fails with container linux stable (2135.6.0) and docker 18.06.3

maybe this was already fixed in a later docker version or it does depend on the underlying os...

**Comment by zyclonite** [comment]
did a build with the following change https://github.com/zyclonite/ZeroTierOne/commit/3dec78f50963d616ca3d469b2e55c0fbf30ec231

you can test it using `docker pull zyclonite/zerotier:1.4.2-3`

works for me now

**Comment by zyclonite** [comment]
@adamierymenko if it looks ok for you, i can do a pr

**Comment by adamierymenko** [comment]
It looks "okay" but I don't understand why it works. What's the nl_pid file used for? Paging @glimberg

**Comment by adamierymenko** [comment]
```
nl_pid is the unicast address of netlink socket.  It's always 0 if
       the destination is in the kernel.  For a user-space process, nl_pid
       is usually the PID of the process owning the destination socket.
       However, nl_pid identifies a netlink socket, not a process.  If a
       process owns several netlink sockets, then nl_pid can be equal to the
       process ID only for at most one socket.  There are two ways to assign
       nl_pid to a netlink socket.  If the application sets nl_pid before
       calling bind(2), then it is up to the application to make sure that
       nl_pid is unique.  If the application sets it to 0, the kernel takes
       care of assigning it.  The kernel assigns the process ID to the first
       netlink socket the process opens and assigns a unique nl_pid to every
       netlink socket that the process subsequently creates.
```

Eww.

**Comment by adamierymenko** [comment]
Okay I did what you did... can you pull `dev` and see if this works?

**Comment by adamierymenko** [comment]
Fixed in `dev`

**Comment by zyclonite** [comment]
@adamierymenko my theory what happens if PID is used

the process in the container sees only his PID which will be 1 but on the host system may be a process (e.g. init) which runs as 1 as well
in the `net=host` situation the host network is shared in the container so i guess it can run into a conflict

thank you for the fix, lgtm

**Comment by zyclonite** [comment]
tested as fixed in 1.4.4
